rm(list=ls())

library(rhdf5)
library(ggplot2)
library(dplyr)

# load riboviz functions
riboviz_rscripts_dirname <- "."
source(file.path(riboviz_rscripts_dirname, "stats_figs_block_functions.R"))
source(file.path(riboviz_rscripts_dirname, "read_count_functions.R"))

# data from YAML read with optparse in generate_stats_figs.R
# asite_disp_length_file <- "../data/yeast_standard_asite_disp_length.txt"
# hd_file <- "../vignette/output/WTnone/WTnone.h5"
# orf_gff_file <- "../vignette/input/yeast_YAL_CDS_w_250utrs.gff3"
# dataset <- "vignette"
hd_file <- "../green/output/sample_1/sample_1.h5"
orf_gff_file <- "../green/input/scer.transcripts.20cds20.gff3"
dataset <- "green"
min_read_length <- 10
max_read_length <- 50

# determine range of valid distances between 5' end and A site
start_codon_max_dist5 <- sapply(min_obs_length:max_obs_length,
                                function(x) {
                                  rpf_length_cts <- subset(start_codon, ReadLen==x)
                                  rpf_length_cts$dist5 <- -rpf_length_cts$Pos + 3
                                  max_dist5 <- sapply(seq(nrow(rpf_length_cts)),
                                                      function(x) {
                                                        any(rpf_length_cts$Counts[1:x]!=0)
                                                      })
                                  max_dist5 <- rpf_length_cts$dist5[which.max(max_dist5)]
                                  return(max_dist5)
                                })


start_codon_plot <- ggplot(start_codon, aes(x=dist5, y=ReadLen, fill=prop)) +
  geom_tile(linetype=1, col=1) + theme_classic() +
  scale_fill_gradient(low="white", high="darkblue") +
  # scale_x_continuous(breaks=seq(min(start_codon$start_dist), max(start_codon$start_dist))) +
  # scale_y_continuous(breaks=fp_min_length:fp_max_length) +
  xlab("position relative to start codon") + ylab("footprint length") +
  theme(axis.text.x=element_text(angle=90, vjust=0.5)) +
  ggtitle("Riboviz vignette data",
          subtitle="start codon footprint density") +
  scale_x_reverse()


# code --------------------------------------------------------------------

gff_df <- readGFFAsDf(orf_gff_file)

# variables to be specified in global environment?
read_range <- min_read_length:max_read_length
nnt_buffer <- 20
nnt_gene <- 0


GenerateAsiteOffsets <- function(dataset, hd_file, gff_df,
                                 min_read_length, max_read_length,
                                 learn_from_data = F,
                                 dist5 = 14:18, dist3 = 9:11,
                                 read_lengths = 27:31) {
  ## return dataframe with columns for read length, frame of read 5' end, A-site offset
  # dataset: character; specified in YAML
  # hd_file: character; filepath to h5 file generated by bam_to_h5.R
  # gff_df: tibble; generated by readGFFAsDf
  # min_read_length: integer; specified in YAML
  # max_read_length: integer; specified in YAML
  # learn_from_data: logical; whether to learn dist5, dist3, and read_lengths from data
  # dist5: integer vector; legal distances between read 5' end and A-site
  # dist3: integer vector; legal distances between read A-site and 3'end (ignored if learn_from_data=T)
  # read_lengths: integer vector; footprint lengths to consider (ignored if learn_from_data=T)

  if(learn_from_data) {
    gene_names <- rhdf5::h5ls(hd_file, recursive = 1)$name
    # 1. determine range of observed read lengths
    read_lengths <- CalculateObservedReadLengths(gene_names, dataset, hd_file,
                                                 min_read_length, max_read_length)
    # 2. determine range of observed distances between read 5' end and A-site
    dist5 <- CalculateObservedDist5()
    # 3. determine range of observed distances from read A-site to 3' end
  } else {
    ## TODO: check that input vectors contain integers
    read_lengths <- sort(read_lengths)
    dist5 <- sort(dist5)
    dist3 <- sort(dist3)
  }
  # 4. assign A-site offsets
  # 5. return output dataframe
  return(read_lengths)

}

CalculateObservedReadLengths <- function(gene_names, dataset, hd_file,
                                         min_read_length=10, max_read_length=50) {
  ## return vector of read lengths with observed counts
  # gene_names: character vector; vector of gene names in gff3 annotation
  # dataset: character; specified in YAML
  # hd_file: character; filepath to h5 file generated by bam_to_h5.R
  # min_read_length: integer; specified in YAML
  # max_read_length: integer; specified in YAML

  # 1. calculate read counts by length
  invisible(capture.output(read_lengths_counts <- CalculateReadLengths(gene_names, dataset, hd_file)))
  # 2. calculate minimum observed read length
  min_obs_length <- sapply(seq(nrow(read_lengths_counts)),
                           function(x) {
                             all(read_lengths_counts$Counts[1:x]==0)
                           })
  min_obs_length <- read_lengths_counts$Length[which.min(min_obs_length)]
  # 3. calculate maximum observed read length
  max_obs_length <- sapply(seq(nrow(read_lengths_counts)),
                           function(x) {
                             all(read_lengths_counts$Counts[(x+1):nrow(read_lengths_counts)]==0)
                           })
  max_obs_length <- read_lengths_counts$Length[which.max(max_obs_length)]
  if(max_obs_length < min_obs_length) { max_obs_length <- max_read_length }
  # 4. return output vector
  return(min_obs_length:max_obs_length)

}

CalculateObservedDist5 <- function(gene_names, dataset, hd_file) {
  ## return vector of observed distances between read 5' end and A-site, according
  ## to start codon read densities
  # gene_names: character vector; vector of gene names in gff3 annotation
  # dataset: character; specified in YAML
  # hd_file: character; filepath to h5 file generated by bam_to_h5.R

  # 1. generate ribogrid upstream of start codon
  start_codon <- AllGenes5StartPositionLengthCountsTibble(gene_names, dataset, hd_file, gff_df)
  ## +3 for start codon in P-site
  ## +1 because Pos=0 corresponds to nucleotide before start codon [based on GetGeneDatamatrix5start()]
  start_codon$dist5 <- -start_codon$Pos + 3 + 1
  # 2. subset to read lengths with observed counts
  observed_lengths <- sort(unique(subset(start_codon, Counts>0)$ReadLen))
  # 3. compute minimum observed distance between read A site adnd 5' end
  # 4. compute maximum observed distance between read A site and 5' end
  max_dist5 <- sapply(observed_lengths,
                      function(len) {
                        rpf_density <- subset(start_codon, ReadLen==len)
                        rpf_density$prop <- rpf_density$Counts / sum(rpf_density$Counts)
                        # return first dist5 where footprint density > 1%
                        if(any(rpf_density$prop > 0.01)) {
                          return(rpf_density$dist5[which.max(rpf_density$prop > 0.005)])
                        } else {
                          return(NA)
                        }
                      })
  # 5. return output vector
  return(min_dist5:max_dist5)

}