# References

JJ Allaire and Yihui Xie and Jonathan McPherson and Javier Luraschi and Kevin Ushey and Aron Atkins and Hadley Wickham and Joe Cheng and Winston Chang and Richard Iannone (2021). rmarkdown: Dynamic Documents for R. R package version 2.7. URL https://rmarkdown.rstudio.com.

Brandl G et al., https://github.com/sphinx-doc/sphinx/

Winston Chang, Joe Cheng, JJ Allaire, Yihui Xie and Jonathan McPherson (2020). shiny: Web Application Framework for R. R package version 1.5.0. https://CRAN.R-project.org/package=shiny.

Cock, P.J.A., Antao, T., Chang, J.T., Chapman, B.A., Cox, C.J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B. and de Hoon, M.J.L. (2009) Biopython: freely available Python tools for computational molecular biology and bioinformatics, Bioinformatics, Volume 25, Issue 11, 1 June 2009, Pages 1422-1423. doi: [10.1093/bioinformatics/btp163](https://doi.org/10.1093/bioinformatics/btp163).

Andrew Collette, Thomas A Caswell, James Tocknell, Thomas Kluyver, Darren Dale, Anthony Scopatz, et al.. (2019, September 6). h5py/h5py: 2.10.0 (Version 2.10.0). Zenodo. doi: [10.5281/zenodo.3401726](http://doi.org/10.5281/zenodo.3401726).

Trevor L Davis (2020). optparse: Command Line Option Parser. R package version 1.6.6. https://CRAN.R-project.org/package=optparse.

Bernd Fischer, Gregoire Pau and Mike Smith (2019). rhdf5: R Interface to HDF5. R package version 2.30.1. https://github.com/grimbough/rhdf5.

gffutils, http://daler.github.io/gffutils/

GraphViz, https://www.graphviz.org/

The HDF Group. Hierarchical Data Format, version 5, 1997-2020, http://www.hdfgroup.org/HDF5/.

Heger, A. et al, pysam, https://github.com/pysam-developers/pysam/.

Jim Hester (2020). glue: Interpreted String Literals. R package version 1.4.1. https://CRAN.R-project.org/package=glue.

Jim Hester, Florent Angly and Russ Hyde (2020). lintr: A 'Linter' for R Code. R package version 2.0.1. https://CRAN.R-project.org/package=lintr.

Jim Hester, Kirill Müller, Kevin Ushey, Hadley Wickham and Winston Chang (2020). withr: Run Code 'With' Temporarily Modified Global State. R package version 2.3.0. https://CRAN.R-project.org/package=withr.

Kim D., Langmead, B. and Salzberg, S.L. (2015) HISAT: a fast spliced aligner with low memory requirements. Nature Methods 2015. doi: [10.1038/nmeth.3317](https://doi.org/10.1038/nmeth.3317).

Krekel H et al., https://github.com/pytest-dev/pytest/

Langmead, B. and Salzberg S. (2012) Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359. doi: [10.1038/nmeth.1923](https://doi.org/10.1038/nmeth.1923).

M. Lawrence, R. Gentleman, V. Carey: "rtracklayer: an {R} package for interfacing with genome browsers". Bioinformatics 25:1841-1842. doi: [10.1093/bioinformatics/btp328](https://doi.org/10.1093/bioinformatics/btp328).

Li et al. (2009) The Sequence Alignment/Map format and SAMtools, Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: [10.1093/bioinformatics/btp352](https://doi.org/10.1093/bioinformatics/btp352).

Martin, M. (2011) Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal, 17(1), pp. 10-12. doi:[10.14806/ej.17.1.200](https://doi.org/10.14806/ej.17.1.200).

M. Morgan, S. Anders, M. Lawrence, P. Aboyoun, H. Pages, and R. Gentleman (2009): "ShortRead: a Bioconductor package for input, quality assessment and exploration of high-throughput sequence data". Bioinformatics 25:2607-2608. doi: [10.1093/bioinformatics/btp450](https://doi.org/10.1093/bioinformatics/btp450).

Martin Morgan (2019). BiocManager: Access the Bioconductor Project Package Repository. R package version 1.30.10. https://CRAN.R-project.org/package=BiocManager.

Martin Morgan, Hervé Pagés, Valerie Obenchain and Nathaniel Hayden (2020). Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import. R package version 2.2.3. http://bioconductor.org/packages/Rsamtools.

Kirill Müller (2020). here: A Simpler Way to Find Your Files. R package version 1.0.1. https://CRAN.R-project.org/package=here.

Kirill Müller and Lorenz Walthert (2020). styler: Non-Invasive Pretty Printing of R Code. R package version 1.3.2. https://CRAN.R-project.org/package=styler.

The pandas development team (2020) pandas-dev/pandas. Zenodo. doi: [10.5281/zenodo.3898987](http://doi.org/10.5281/zenodo.3898987).

Pandoc, https://pandoc.org/

pigz, http://zlib.net/pigz/

pycodestyle, https://github.com/pycqa/pycodestyle/

pylint, https://www.pylint.org/

pyyaml, https://pyyaml.org/

Quinlan, A.R. and Hall I.M. (2010) BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841-842. doi: [10.1093/bioinformatics/btq033](https://doi.org/10.1093/bioinformatics/btq033).

R Core Team (2021) R: A Language and Environment for Statistical Computing, R Foundation for Statistical Computing, Vienna, Austria, 2021. https://www.R-project.org.

Schlaich M et al., https://github.com/pytest-dev/pytest-cov/

Shirley, M.D., Ma, Z., Pedersen, B., Wheelan, S. (2015) Efficient "pythonic" access to FASTA files using pyfaidx. PeerJ PrePrints 3:e1196. 2015. doi: [10.7287/peerj.preprints.970](https://doi.org/10.7287/peerj.preprints.970).

C. Sievert. Interactive Web-Based Data Visualization with R, plotly, and shiny. Chapman and Hall/CRC Florida, 2020. ISBN: 978-1138331495.

Smith, T.S., Heger , A. and Sudbery, I. (2017) UMI-tools: Modelling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome Res. 2017, January 18 2017, doi: [10.1101/gr.209601.116](https://doi.org/10.1101/gr.209601.116).

Di Tommaso, P., Chatzou, M., Floden, E. W., Barja, P. P., Palumbo, E., & Notredame, C. (2017). Nextflow enables reproducible computational workflows. Nature Biotechnology, 35(4), 316-319. doi: [10.1038/nbt.3820](https://doi.org/10.1038/nbt.3820).

Trier M et al., https://github.com/gitpython-developers/GitPython/

Kevin Ushey (2018). RcppRoll: Efficient Rolling / Windowed Operations. R package version 0.3.0. https://CRAN.R-project.org/package=RcppRoll.

Hadley Wickham. testthat: Get Started with Testing. The R Journal, vol. 3, no. 1, pp. 5-10, 2011. https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf.

Wickham et al., (2019). Welcome to the tidyverse. Journal of Open  Source Software, 4(43), 1686. doi: [10.21105/joss.01686](https://doi.org/10.21105/joss.01686).

Hadley Wickham, Peter Danenberg, Gábor Csárdi and Manuel Eugster (2020). roxygen2: In-Line Documentation for R. R package version 7.1.1. https://CRAN.R-project.org/package=roxygen2.

Hadley Wickham, Jim Hester and Winston Chang (2020). devtools: Tools to Make Developing R Packages Easier. R package version 2.3.2. https://CRAN.R-project.org/package=devtools.

Widgren, S., et al. (2020) git2r: Provides Access to Git Repositories. R package version 0.27.1. URL https://CRAN.R-project.org/package=git2r.

Yihui Xie (2014) knitr: A Comprehensive Tool for Reproducible Research in R. In Victoria Stodden, Friedrich Leisch and Roger D. Peng, editors, Implementing Reproducible Computational Research. Chapman and Hall/CRC. ISBN 978-1466561595.

Yihui Xie (2015) Dynamic Documents with R and knitr. 2nd edition. Chapman and Hall/CRC. ISBN 978-1498716963.

Yihui Xie and J.J. Allaire and Garrett Grolemund (2018). R Markdown: The Definitive Guide. Chapman and Hall/CRC. ISBN 9781138359338. URL https://bookdown.org/yihui/rmarkdown.

Yihui Xie and Christophe Dervieux and Emily Riederer (2020). R Markdown Cookbook. Chapman and Hall/CRC. ISBN 9780367563837. URL https://bookdown.org/yihui/rmarkdown-cookbook.

Yihui Xie (2021). knitr: A General-Purpose Package for Dynamic Report Generation in R. R package version 1.33. https://CRAN.R-project.org/package=knitr.
